android/app/src/main/java/com/zulipmobile/notifications/GCMPushNotifications.java

```diff
@@ -0,0 +1,94 @@
+import com.wix.reactnativenotifications.core.AppLifecycleFacade;
+import com.wix.reactnativenotifications.core.JsIOHelper;
+import com.wix.reactnativenotifications.core.notification.PushNotification;
+import com.zulipmobile.R;
+
+import java.io.IOException;
+import java.net.URL;
+
+public class GCMPushNotifications extends PushNotification {
+
+    public GCMPushNotifications(Context context, Bundle bundle, AppLifecycleFacade appLifecycleFacade, AppLaunchHelper appLaunchHelper, JsIOHelper jsIoHelper) {
+        super(context, bundle, appLifecycleFacade, appLaunchHelper, jsIoHelper);
+    }
+
+    @Override
+    protected PushNotificationsProp createProps(Bundle bundle) {
+        return new PushNotificationsProp(bundle);
+    }
+
+    protected PushNotificationsProp getProps() {
+        return (PushNotificationsProp) mNotificationProps;
+    }
+
+    @Override
+    protected Notification.Builder getNotificationBuilder(PendingIntent intent) {
+        // First, get a builder initialized with defaults from the core class.
+        final Notification.Builder builder = super.getNotificationBuilder(intent);
+
+        String type = getProps().getRecipientType();
+        String content = getProps().getContent();
+        String title = getProps().getSenderFullName();
+        String avatarURL = getProps().getAvatarURL();
+        String time = getProps().getTime();
+        String stream = getProps().getStream();
+        String topic = getProps().getTopic();
+        String baseURL = getProps().getBaseURL();
+
+
+        builder.setSmallIcon(R.drawable.zulip_notification);
+        builder.setContentTitle(title);
+        builder.setAutoCancel(true);
+        builder.setContentText(content);
+
+        if (type.equals("private")) {
+            if (android.os.Build.VERSION.SDK_INT >= 16) {
+                builder.setSubText("New private message");
+            }
+        } else if (type.equals("stream")) {
+            if (android.os.Build.VERSION.SDK_INT >= 16) {
+                String displayTopic = stream + " > "
+                        + topic;
+                builder.setSubText("Mention on " + displayTopic);
+            }
+        }
+
+        if (avatarURL != null) {
+            Bitmap avatar = fetchAvatar(NotificationHelper.sizedURL(mContext,
+                    avatarURL, 64, baseURL));
+            if (avatar != null) {
+                builder.setLargeIcon(avatar);
+            }
+        }
+
+        if (time != null) {
+            long timStamp = Long.parseLong(getProps().getTime()) * 1000;
+            builder.setWhen(timStamp);
+        }
+
+        long[] vPattern = {0, 100, 200, 100};
+        builder.setVibrate(vPattern);
+        return builder;
+    }
+
+
+    private Bitmap fetchAvatar(URL url) {
+        try {
+            return NotificationHelper.fetch(url);
+        } catch (IOException e) {
+            Log.e("ERROR", e.toString());
+        }
+        return null;
+    }
+}
```

Comment:
So the problem is, the library has a service running which receives the GCM notification and displays the notification it self, hence there is no js code manipulating the notification data! 
Therefore I'm facing a problem with avatar as the server does not sends the full URL of the avatar if its uploaded (/user_uploads/....) and currently crashing.
So most probably I'll have to modify the server part.

Correct me if I'm wrong